cli/azd/internal/vsrpc/handler.go

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package vsrpc
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"go.lsp.dev/jsonrpc2"
+)
+
+// Handler is the type of function that handles incoming RPC requests.
+type Handler func(ctx context.Context, conn jsonrpc2.Conn, reply jsonrpc2.Replier, req jsonrpc2.Request) error
+
+// HandlerAction0 is a helper for creating a Handler from a function that takes no arguments and returns an error.
+func HandlerAction0(f func(context.Context) error) Handler {
+	return func(ctx context.Context, conn jsonrpc2.Conn, reply jsonrpc2.Replier, req jsonrpc2.Request) error {
+		err := f(ctx)
+		return reply(ctx, nil, err)
+	}
+}
+
+// HandlerAction1 is a helper for creating a Handler from a function that takes one argument and returns an error.
+func HandlerAction1[T1 any](f func(context.Context, T1) error) Handler {
+	return func(ctx context.Context, conn jsonrpc2.Conn, reply jsonrpc2.Replier, req jsonrpc2.Request) error {
+		t1, err := unmarshalArg[T1](req, 0)
+		if err != nil {
+			return reply(ctx, nil, err)
+		}
+
+		err = f(ctx, t1)
+		return reply(ctx, nil, err)
+	}
+}
+
+// HandlerFunc0 is a helper for creating a Handler from a function that takes no arguments and returns a value and an error.
+func HandlerFunc0[TRet any](f func(context.Context) (TRet, error)) Handler {
+	return func(ctx context.Context, conn jsonrpc2.Conn, reply jsonrpc2.Replier, req jsonrpc2.Request) error {
+		ret, err := f(ctx)
+		return reply(ctx, ret, err)
+	}
+}
+
+// HandlerFunc1 is a helper for creating a Handler from a function that takes one argument and returns a value and an error.
+func HandlerFunc1[T1 any, TRet any](f func(context.Context, T1) (TRet, error)) Handler {
+	return func(ctx context.Context, conn jsonrpc2.Conn, reply jsonrpc2.Replier, req jsonrpc2.Request) error {
+		t1, err := unmarshalArg[T1](req, 0)
+		if err != nil {
+			return reply(ctx, nil, err)
+		}
+
+		ret, err := f(ctx, t1)
+		return reply(ctx, ret, err)
+	}
+}
```

Comment:
Just thinking if there is a more generic way we can handle the different type of RPC service signatures as it relates to input arguments and return values.  

I don't have anything specific in mind at the moment.

Reply:
I think there might be a way to do this using go's reflection APIs, i.e. `reflect.Method` and friends - but I didn't want to nerd snipe myself trying to write gnarly generics + reflection code (which would involve learning a lot more about go generics and reflection than I already knew) to try to make it a little cleaner right now as I was just trying to get something working, and the set of these handlers was small enough that I could just write them all by hand.

But yeah, kinda gross and I wish there was something a little more dynamic we can do - there may be via reflection...

Reply:
I looked at some of the implementation of the IOC Container we build atop of to see how it used reflection and if a similar pattern could be used here - I'm pretty confident that if we wanted we can add some:

`HandlerForMethod(method interface{}) Handler`

Which would use reflection to ensure that method was a Func and then looked at it's arguments and built the correct wrapper.

But I am also pretty sure that for the short term I am just going to continue to hand roll these helpers since the set of function signatures is pretty small (only varies in argument arity and return arity) and that feels better than writing a bunch of reflection code and frees our time up to work on other stuff.